### ISSUE TITLE
Exploring adapters transformers for different oneQA tasks

### ISSUE THREAD

**Issue by jdpsen** [primeqa/primeqa#176]
_jaydesen created the following on Feb 08:_  

 The aim is to see if adapter transformers can achieve similar/better performance compared to regular BERT/Transformer architectures. The hope is to keep same BERT/Transformer layer unchanged underneath different oneQA tasks and use adapter layers for task-based finetuning. 

_Original issue: https://github.ibm.com/ai-foundation/PrimeQA/issues/23_

**Comment by jdpsen** [comment]
_On Mar 02, jsmc commented:_
Adapter experiments on list questions from NQ:
This is a small set,  2713 train, 73 test
First three lines (baselines, no adapters) previous reported by Sara, [](https://ibm.ent.box.com/file/885572310095?s=oi5c5cxpm1uznxesgabqmlcvut06iio5)

Models with tydi have been training on tydi for MRC.  Models with xlmr have only seen MLM pretraining, no MRC data

1: tydi a tydi model, no additional training on NQ
2: xlmr tuned only NQ lists
3: tydi tuned only on NQ lists
4,5: xlmr with an adapter that sees NQ lists, compare with 2
6,7: tydi with an adapter that sees NQ lists, compare with 3

Adapter models are averaged over 9 seeds, after learning rate sweep (5e-4 chosen)

| model | passage mean | epochs | stdev | minimal | stdev |
|-------|----------------|--------|-------|---------|-------|
| tydi, no tuning | | 32.1 | | 17.1 | |
| xlmr + tuning(NQ) | | 49.7 | | 48.8 | |
| tydi + tuning(NQ) || 56.1 || 58.1 ||
| xlmr+adapter(NQ) | 1 | 46.83 | 0.86 | 45.48 | 1.20 |
| xlmr+adapter(NQ) | 10 | 53.58 | 2.67 | 52.67 | 2.16 |
| tydi+adapter(NQ) | 1 | 56.53 | 2.71 | 61.60 | 1.44 |
| tydi+adapter(NQ) | 10 | 54.52 | 2.18 | 60.86 | 1.43 |

(1) adapter-based models have similar to slightly better performance than comparable non-adapter models (with above caveat about learning rate sweep)
(2) using xlmr as a starting point performs considerably better after 10 epochs
(3) using tydi as starting point, 1 epoch works fine, seems overtrained after 10 epochs
(4) variance is very large due small test set size

**Comment by jdpsen** [comment]
_On Mar 08, jaydesen commented:_
Experiments in progress with GooAQ. - initial observation shows not so large variance as in listNQ.

**Comment by jdpsen** [comment]
_On Mar 14, jsmc commented:_





 
Results on GooAQ test set.
As before, adapter performance is comparable to non-adapter baselines:


| model                  |   passage |   minimum |
|:-----------------------|----------:|----------:|
| non-adapter baseline   |     85.8  |     83    |
| xlmr+adapter(NQ)       |     85.99 |     82.69 |
| non-adapter baseline * |     83.7  |     86.6  |
| Tydi+adapter(NQ)       |     86.06 |     82.9  |

(*) I wonder if there is a swap in the spreadsheet.




 Learning curves for adapters on the NQ dataset.
1 files is about 1300 events.
Results on file 15 held-out from training set.

| init   | # train file   |   epochs | passage        | minimal        | av        |
|:-------|:---------------|---------:|:---------------|:---------------|:----------|
| tydi   | 300 examples   |       10 | 62.90 +/- 0.15 | 38.50 +/- 0.13 | 7 subsets |
| tydi   | 600 examples   |       10 | 65.60 +/- 0.37 | 44.08 +/- 0.40 | 7 subsets |
| tydi   | 1              |        1 | 78.40 +/- 0.49 | 65.98 +/- 0.45 | 7 subsets |
| tydi   | 1              |        5 | 79.47 +/- 0.45 | 71.81 +/- 0.81 | 7 subsets |
| tydi   | 1              |       10 | 79.51 +/- 0.79 | 72.23 +/- 0.76 | 7 subsets |
| tydi   | 2              |       10 | 80.89 +/- 1.00 | 74.57 +/- 1.24 | 3 subsets |
| tydi   | 4              |       10 | 83.02 +/- 0.82 | 78.36 +/- 0.81 | 3 subsets |
| tydi   | 8              |       10 | 84.34 +/- 0.57 | 81.06 +/- 0.35 | 3 subsets |
| tydi   | 15             |       10 | 85.80 +/- 0.24 | 82.80 +/- 0.35 | 7 seeds   |
| xlmr   | 1              |        1 | 61.84 +/- 0.69 | 44.29 +/- 1.35 | 7 subsets |
| xlmr   | 1              |        5 | 72.40 +/- 1.15 | 61.52 +/- 1.23 | 7 subsets |
| xlmr   | 1              |       10 | 73.49 +/- 1.18 | 63.47 +/- 1.53 | 7 subsets |
| xlmr   | 2              |       10 | 77.80 +/- 0.31 | 69.57 +/- 0.12 | 3 subsets |
| xlmr   | 4              |       10 | 80.99 +/- 1.07 | 75.16 +/- 0.53 | 3 subsets |
| xlmr   | 8              |       10 | 83.22 +/- 0.68 | 78.93 +/- 0.07 | 3 subsets |
| xlmr   | 15             |       10 | 85.00 +/- 0.39 | 74.80 +/- 0.56 | 7 seeds   |

(4) advantage of starting from tydi is much greater with less data
(5) more epochs are needed with xlmr init, limited data, whereas Tydi init is good at one epoch
(6) more variance with xlmr init and limited data??

**Comment by jdpsen** [comment]
_On Mar 24, jaydesen commented:_
NQ might be a good start for Macy's

**Comment by jdpsen** [comment]
_On Mar 29, jaydesen commented:_
Macy's data testing and booleanQA debugging in progress .
The bug was in GAAMA MRC pipeline - answer misses the start index -- typically when the answer is the beginning of the doc. GRPC works that way.

**Comment by jdpsen** [comment]
_On Apr 05, jaydesen commented:_
What if question words are omitted ? e.g. Wh words getting dropped might cause issue in classifier. Can adapter adapt to that  . Expt are in progress.

**Comment by jdpsen** [comment]
@jsmcibm : You can update your current progress on this issue.

**Comment by jsmcibm** [comment]
Immediate objectives
(1) test harness to benchmark document-at-a-time modeling, using adapter-based models equivalent to the boolean pipeline.
(2) build a balanced dataset, with a reasonably balanced mixture of factoid, boolean, list, and domain-specific questions
(3) understand tokenizer overhead issue seen in profiling

**Comment by jsmcibm** [comment]
Additional tasks, beyond boolean question type classification and boolean evidence classification:
(1) domain adaptation - Covid or Macy's
(2) List questions
